### DISCUSSION TITLE
Using ArchiveBox with git-annex

### DISCUSSION THREAD

**Discussion by hatzka-nezumi** [ArchiveBox/ArchiveBox#1592] [Q&A] [answered]
I have a collection about 18 TiB in size, mostly of YouTube videos. It's managed by a pile of homegrown scripts that have become too messy to handle, and I just came across ArchiveBox, so I'm thinking of migrating to it. I'll worry about the reorganization myself (most likely I'll just extract the URLs from the current collection, redownload everything that still exists publicly, and move anything that doesn't into a separate manually-managed section), but I also use and like git-annex, and I'm wondering if that will pose any problems for ArchiveBox.

If you aren't aware, [git-annex](https://git-annex.branchable.com) is a tool that finds large files in a git repository, hashes their contents, moves them into a subfolder of `.git` using the hash as the filename, and finally replaces the original file with a symlink. git only sees the symlink, so it doesn't have to process the whole file, but since the destination's name is a hash, you still know if you have the right content. In addition, you can spread the collection files across multiple clones of the repository; in this case, no single clone needs to have all of the files, but every clone is kept up-to-date on which files are where (because this is tracked in another git branch).

The main issues that I can imagine popping up are:

- ArchiveBox will need to be taught to commit to the repository after each job runs. I can probably do this myself.
- If a file isn't present locally (to a given checkout), it shows up as a broken symlink. If ArchiveBox itself checks for the existence of such files, it might get confused. This seems like an easy issue to fix (famous last words), as it only necessitates replacing regular file-exists checks with `lstat` in git-annex contexts.
- git-annex is optimized for the case where the large files change rarely. This is true of the archived data itself, which should never change, but the SQLite database can't be handled this way. (I mean, it _could_ be, but then I'd be storing a separate copy of the entire database for every change...)

But I don't know ArchiveBox's internals well enough to say for sure if these are the only issues. Is anyone else doing this? Has anyone tried and failed?

**Comment by pirate** [comment]
I would not recommend migrating that collection to ArchiveBox for a couple reasons:

- it doesn't yet provide any native deduping/compression options, Archivebox assumes the underlying filesystem (e.g. ZFS) will take care of that
- it's not primarily designed for massive mostly-video libraries, the UI to play and manage videos is not that nice compared to something like jellyfin/plex/etc.
- it doesn't provide any organization features other than flat tags, so you can't folderize things in a hierarchy or take notes on entries within ArchiveBox

> I'll just extract the URLs from the current collection, redownload everything that still exists publicly, and move anything that doesn't...

I think this may be more error-prone than you're expecting. there isn't a great way to tell if something has redownloaded "properly", as often requests will respond with 200 OK but return a page that has a modal/popup/ad/CAPTCHA/login window hiding the content. Don't be so quick to throw out your original copies just because it looks like a new tool has saved the URLs at first glance. I have built an LLM-powered QA system to detect issues like that in captures (for one of our paying clients), but it's not yet publically available. (WebRecorder.net also provides good QA tooling to assess capture success rate.)


If you do decide to go for it, definitely try it at a small scale first with ~1,000 videos and see if it works well for your needs first.

**Reply by hatzka-nezumi** [reply]
> it doesn't yet provide any native deduping/compression options, Archivebox assumes the underlying filesystem (e.g. ZFS) will take care of that

git-annex provides file-level deduplication of annexed files, since their real names behind the symlinks are hashes of their content. I don't currently have anything more than that, and I wasn't expecting it from ArchiveBox; this kind of thing shouldn't be its focus, IMO.

> it's not primarily designed for massive mostly-video libraries, the UI to play and manage videos is not that nice compared to something like jellyfin/plex/etc.

I'm not coming from Jellyfin or Plex. I'm not coming from anything, really; when I started out, I decided to worry first about getting things into the collection, not making them easy to access, since that part can be done later, whereas once something is off the internet you can't get it back. ATM the UI is "grep through the info files for the keywords you want, then manually open the video file in mpv". ArchiveBox can't be worse than that.

In the long run, I figure I can rig up a script that reads the yt-dlp metadata and generates a friendlier folder structure for Jellyfin or whatever, sharing the actual media files with the ArchiveBox dataset through more symlinks to save disk space. That was my original plan without ArchiveBox, anyway. Really, I'd mostly be using it as a downloader, not as the main way of accessing the collection.

> it doesn't provide any organization features other than flat tags, so you can't folderize things in a hierarchy or take notes on entries within ArchiveBox

Again, not something I currently have, and not something I really expected out of ArchiveBox.

> I think this may be more error-prone than you're expecting. there isn't a great way to tell if something has redownloaded "properly", as often requests will respond with 200 OK but return a page that has a modal/popup/ad/CAPTCHA/login window hiding the content. 

This is a good point, but it should be possible to write a script to do a few sanity checks, like confirming that there's a video file, that its metadata matches, that the quality is as good as my current copy, and so on. It'll be work, but I don't see any reason it's not doable. And anyway, it's not as though I've manually checked that each and every video in the current collection downloaded properly either.

**Reply by pirate** [reply, marked as answer]
Ok, makes sense, just wanted to make sure you're aware of the limitations for this use case.

As for your specific concerns:

> ArchiveBox will need to be taught to commit to the repository after each job runs. I can probably do this myself.

Yeah I would do this out-of-band with a periodic cronjob or inotify watcher.

> If a file isn't present locally (to a given checkout), it shows up as a broken symlink. If ArchiveBox itself checks for the existence of such files, it might get confused. 

ArchiveBox exclusively uses `os.path.isfile(...)` for file existence, which returns `False` for broken symlinks. However, ArchiveBox wont try to re-download files once a success is recorded in the DB, even if it sees the files on disk go missing. So I don't think git-annex will cause any problems here.

> git-annex is optimized for the case where the large files change rarely. This is true of the archived data itself, which should never change, but the SQLite database can't be handled this way.

Most archivebox data is append-only/doesn't change often, however the `data/archive/<ts>/index.{html,json}` files do get mutated a few times during the archiving process as each extractor runs. 

As for the SQLite database, you could exclude `index.sqlite3` in your `.gitignore` but have a cronjob that copies it to `index.sqlite3.bak` (a non-excluded path) every 24hr. That way you still have all your data in one place in git, but with DB commits rate-limited to 24hr so it wouldn't use up as much storage space.
